File: scripts/init-rec-mint.ts
// Initialize the fungible REC mint (governance program).
//
// REC is a Token-2022 SPL mint: 1 token = 1 MWh, 6 decimals (base unit = 1 Wh, so
// 1 kWh = 1_000 base units). PDA = [b"rec_mint"]; mint authority = the governance
// [b"poa_config"] PDA. `issue_erc` mints REC to producers; `retire_rec` burns them.
//
// Run AFTER init-governance.ts (the poa_config PDA must already exist):
//   ANCHOR_PROVIDER_URL=http://localhost:8899 ANCHOR_WALLET=<dev-wallet> npx tsx scripts/init-rec-mint.ts
import * as anchor from '@anchor-lang/core';
import { Program } from '@anchor-lang/core';
import { Governance } from '../target/types/governance';
import { PublicKey, SystemProgram } from '@solana/web3.js';
import { TOKEN_2022_PROGRAM_ID } from '@solana/spl-token';

async function main() {
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);
  const governance = anchor.workspace.Governance as Program<Governance>;
  const authority = provider.wallet.publicKey;

  const [poaPda] = PublicKey.findProgramAddressSync([Buffer.from('poa_config')], governance.programId);
  const [recMint] = PublicKey.findProgramAddressSync([Buffer.from('rec_mint')], governance.programId);

  console.log('REC mint PDA:', recMint.toBase58());
  try {
    const sig = await governance.methods
      .initRecMint()
      .accounts({
        governanceConfig: poaPda,
        recMint,
        authority,
        tokenProgram: TOKEN_2022_PROGRAM_ID,
        systemProgram: SystemProgram.programId,
      } as any)
      .rpc();
    console.log('  ✅ REC mint initialized (6 decimals, authority = poa_config PDA):', sig);
  } catch (e: any) {
    if (/already in use|0x0/.test(e.message ?? '')) {
      console.log('  ℹ️ REC mint already initialized');
    } else {
      throw e;
    }
  }
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: programs/governance/src/instructions/init_rec_mint.rs
use crate::errors::GovernanceError;
use crate::state::*;
use anchor_lang::prelude::*;
use anchor_spl::token_interface::{Mint as MintInterface, TokenInterface};

#[derive(Accounts)]
pub struct InitRecMint<'info> {
    #[account(
        seeds = [b"governance_config"],
        bump,
        has_one = authority @ GovernanceError::UnauthorizedAuthority
    )]
    pub governance_config: Account<'info, GovernanceConfig>,
    /// Fungible REC mint: 1 token = 1 MWh, 6 decimals (base unit = 1 Wh; 1 kWh = 1000 units).
    /// Mint authority = governance_config PDA `[b"governance_config"]`.
    #[account(
        init,
        payer = authority,
        seeds = [b"rec_mint"],
        bump,
        mint::decimals = 6,
        mint::authority = governance_config,
        mint::token_program = token_program,
    )]
    pub rec_mint: Box<InterfaceAccount<'info, MintInterface>>,
    #[account(mut)]
    pub authority: Signer<'info>,
    pub token_program: Interface<'info, TokenInterface>,
    pub system_program: Program<'info, System>,
}

/// Initialize the fungible REC mint (PDA `[b"rec_mint"]`, 6 decimals, mint authority =
/// governance_config PDA). Run once before issuing certificates. The mint is created
/// entirely via account constraints.
pub fn init_rec_mint(_ctx: Context<InitRecMint>) -> Result<()> {
    Ok(())
}
